File: cpp/cpp_100-999/309.cc
// C++
// leetcode 309
// https://leetcode-cn.com/problems/best-time-to-buy-and-sell-stock-with-cooldown/

class Solution {
public:
    int maxProfit(vector<int>& prices) {
    	int buy = INT_MIN;
    	int sell = 0;
    	int secondSell = 0;
    	bool cooling = false;
    	for(auto & x: prices) {
    		if(!cooling) {
    			buy = max(buy, sell - x);
    		} else {
    			buy = max(buy, secondSell - x);
    		}
    		if(buy + x > sell) {
    			secondSell = sell;
    			sell = buy + x;
    			cooling = true;
    		} else
    			cooling = false;
    	}
    	return sell;
    }
};

File: Go/Go_100-999/309.go
// Go
// leetcode 309
// https://leetcode.cn/problems/best-time-to-buy-and-sell-stock-with-cooldown/

func maxProfit(prices []int) int {
	cnt := []int{-10000000, 0, 0} // 持有，可买，冻结
	for _, x := range prices {
		cnt[0] = max(cnt[0], cnt[1]-x)
		cnt[1] = max(cnt[1], cnt[2])
		cnt[2] = cnt[0] + x
	}
	return max(cnt[1], cnt[2])
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}